litmus-portal/backend/self-deployer/pkg/k8s/cluster.go

```diff
@@ -0,0 +1,95 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/client-go/discovery"
+	memory "k8s.io/client-go/discovery/cached"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+var decUnstructured = yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+
+func applyRequest(obj *unstructured.Unstructured, requestType interface{}, dr dynamic.ResourceInterface) (*unstructured.Unstructured, error) {
+
+	if requestType == "create" {
+		response, err := dr.Create(context.TODO(), obj, metav1.CreateOptions{})
+		if err != nil {
+			log.Printf("err: %v\n", err)
+		}
+		return response, nil
+	} else if requestType == "update" {
+		response, err := dr.Update(context.TODO(), obj, metav1.UpdateOptions{})
+		if err != nil {
+			log.Printf("err: %v\n", err)
+		}
+		return response, nil
+	} else if requestType == "delete" {
+		err := dr.Delete(context.TODO(), obj.GetName(), metav1.DeleteOptions{})
+		if err != nil {
+			log.Printf("err: %v\n", err)
+		}
+		return &unstructured.Unstructured{}, nil
+	} else if requestType == "get" {
+		response, err := dr.Get(context.TODO(), obj.GetName(), metav1.GetOptions{})
+		if err != nil {
+			log.Printf("err: %v\n", err)
+		}
+		return response, nil
+	} else {
+		log.Println("Invalid Request")
+		return &unstructured.Unstructured{}, fmt.Errorf("Invalid request")
+	}
+}
+
+func ClusterOperations(manifest []byte, requestType string) (*unstructured.Unstructured, error) {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		log.Printf("err: %v\n", err)
+	}
+
+	dc, err := discovery.NewDiscoveryClientForConfig(cfg)
+	if err != nil {
+		log.Printf("err: %v\n", err)
+	}
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(dc))
+
+	// Prepare the dynamic client
+	dyn, err := dynamic.NewForConfig(cfg)
+	if err != nil {
+		log.Printf("err: %v\n", err)
+	}
+
+	// Decode YAML manifest into unstructured.Unstructured
+	obj := &unstructured.Unstructured{}
+	_, gvk, err := decUnstructured.Decode([]byte(manifest), nil, obj)
+	if err != nil {
+		log.Printf("err: %v\n", err)
+	}
+
+	// Find GVR
+	mapping, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		log.Printf("err: %v\n", err)
+	}
+
+	// Obtain REST interface for the GVR
+	var dr dynamic.ResourceInterface
+	if mapping.Scope.Name() == meta.RESTScopeNameNamespace {
+		// namespaced resources should specify the namespace
+		dr = dyn.Resource(mapping.Resource).Namespace(obj.GetNamespace())
+	} else {
+		// for cluster-wide resources
+		dr = dyn.Resource(mapping.Resource)
+	}
+
+	return applyRequest(obj, requestType, dr)
+}
```

Comment:
As these are redundant functions from the agent you can import the package using 
`
go get -v github.com/litmuschaos/litmus/litmus-portal/backend/subscriber/agent/pkg/cluster@litmus-portal-v1
`

Reuse those function.

Reply:
Yes I tried, bch is giving warnings but I cannot do that unless it is tagged and published on github first.

Reply:
Cannot reuse the function as my one is different from this one, I am getting the data as an array of bytes directly, I can maybe reuse the applyRequest but isn't exported.

Reply:
You have to use ClusterOperation function
Cluster operation function will call applyRequesr

Reply:
You have to send the data in this format
{
   requesttype: "create",
   manifest: "",
   resourcettype: "",
  namespace: ""
}

Reply:
Yes we can maybe reuse this package in go-agent, this function is more generic and it might be easier to use?

Reply:
Yes